0002-ether-htlc-dynamic-final-addresses.md

```diff
@@ -0,0 +1,121 @@
+Bitcoin's UTXO model means you specify which identity will finally own the funds in the redeem and refund transactions.
+In the Ethereum HTLC this is embedded in the contract from the beginning.
+
+Both contracts are representative of the paradigms of each blockchain but their differences get in the way writing generic code for interacting with them.
+The nature of Bitcoin allows the COMIT node to handle all the keys involved in the HTLC and simply ask the user for where they want the funds when the time comes.
+To me this is good design and what we want in COMIT even outside of atomic swap protocols:
+
+1. The users negotiate a protocol with no keys required from the users' wallets
+2. The protocol execution starts with users funding the initial contracts from their wallets
+3. The COMIT nodes execute the protocol (via the user's instruction and/or automatically).
+4. The protocol execution completes by the users' getting the final assets they own into their wallets.
+
+As a rule of thumb, the COMIT node should own the private keys of whatever identities are embedded in the contracts on the blockchains.
+
+Since Ethereum is flexible we can restructure it to mimic how Bitcoin works and fit it into the above flow.
+
+## Research
+
+Bitcoin requires you witness the HTLC output with the secret and a signature (or a just the refund signature after a timeout).
+The transaction that has the witness will specify where to redeem the funds in its output.
+
+To follow the same pattern in Ethereum, our redeem transaction will need in its data:
+
+1. The secret
+2. The address to give the money to (the final redeem address)
+3. a signature under the redeem identity from the COMIT node
+
+The refund would be the same without (1)
+
+### The Final Address
+
+The decision we have to make here is whether the final address is:
+
+1. Provided in the data
+2. The sender address
+
+(2) is nice because it means that there is no way to mess it up and send a transaction to an address you don't own. It also saves a bit of gas.
+
+On the other hand, that might be a feature.
+You may want to send the asset in the HTLC to a contract or something, so I'm leaning towards (1).
+
+
+### The signature
+
+The key things when deciding what to sign are:
+
+1. It must not be possible to front run the transaction, learn the secret, and then create another transaction that redeems to a different address.
+2. It must not be possible to re-use the signature on a different chain or a different contract on the same chain.
+
+To ensure (1) we need to sign the final address (redeem or refund) with the private key of the corresponding identity.
+To ensure (2) we should sign the HTLC's contract address, the network id and chain id.
+
+So I propose that we sign the following messages for each outcome:
+
+- Redeem: `network_id || chain_id || htlc_contract_address || final_redeem_address`
+- Refund: `network_id || chain_id || htlc_contract_address || final_refund_address`
```

Comment:
I shortly mentioned it on the ticket but let's discuss here: 

We will need to be able generate the same message inside the contract (and hash it) as as outside of the contract, otherwise the caller can pass an arbitrary hash. 

Since there is no opcode for retrieving the `network_id` or `chain_id` we would need to hardcode it into the contract. This would make using these values in the message for a replay attack (in case of a fork) rather useless, as they will always be the same. 

Instead, we should limit the message to be signed to something which is unique to the current swap, e.g. a combination of the `hashed_secret` + the `contract_address` and require the message to be signed from the `redeem_identity` (or `refund_identity` respectively). 

This would not protect us from a replay attack in case of a fork (I probably wouldn't call it attack, it's just a replay with the same behavior), but then again, we want to mimic Bitcoin and depending on the fork, I'd expect the same behavior.

Reply:
IMO the best course of action when making these decisions is to simply guarantee that the TX can never be reused. For that you need the `network_id` and `chain_id` embedded in the contract.

Like you said, you could sign the `secret_hash` instead of these so a replay attack would only work against the same contract on a different chain if it has the same secret AND the same redeem or refund identity (which would be pretty strange). But rather than thinking about how we could justify this never happening, my preference would be to make it impossible.